osu.Game.Tests/Online/TestMultiplayerMessagePackSerialization.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using MessagePack;
+using NUnit.Framework;
+using osu.Game.Online;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.MatchTypes.TeamVersus;
+
+namespace osu.Game.Tests.Online
+{
+    [TestFixture]
+    public class TestMultiplayerMessagePackSerialization
+    {
+        [Test]
+        public void TestSerialiseRoom()
+        {
+            var room = new MultiplayerRoom(1)
+            {
+                MatchState = new TeamVersusRoomState()
+            };
+
+            var serialized = MessagePackSerializer.Serialize(room);
+
+            var deserialized = MessagePackSerializer.Deserialize<MultiplayerRoom>(serialized);
+
+            Assert.IsTrue(deserialized.MatchState is TeamVersusRoomState);
+        }
+
+        [Test]
+        public void TestSerialiseUserStateExpected()
+        {
+            var state = new TeamVersusUserState();
+
+            var serialized = MessagePackSerializer.Serialize(typeof(MatchUserState), state);
+            var deserialized = MessagePackSerializer.Deserialize<MatchUserState>(serialized);
+
+            Assert.IsTrue(deserialized is TeamVersusUserState);
+        }
+
+        [Test]
+        public void TestSerialiseUnionFailsWithSingalR()
```

Comment:
If SignalR is ever fixed to resolve the issue, this test should start passing, notifying us that we can remove the workaround.

Reply:
I'm doubly confused here:

* On a nitpicky note, you mean failing, right?
* On a less nitpicky note, as far as I can see this is testing *only messagepack code paths*. In other words, the only way this could possibly start failing is if messagepack bends over backwards for signalr and starts serialising derived types as base, which seems like pretty broken behaviour for a serialisation library in general. For this to exhibit the properties you want, you'd want to run it through the signalr wringer that loses the base type, which is likely infeasible. May still be fine to keep the test for illustrative purposes, but if this ever gets fixed on the signalr side, this test won't let you know 99%.

Reply:
> I'm doubly confused here:
> 
> * On a nitpicky note, you mean failing, right?

i did yes.

> * On a less nitpicky note, as far as I can see this is testing _only messagepack code paths_. In other words, the only way this could possibly start failing is if messagepack bends over backwards for signalr and starts serialising derived types as base despite the explicit derived type spec, which seems like pretty broken behaviour for a serialisation library in general. For this to exhibit the properties you want, you'd want to run it through the signalr wringer that loses the base type, which is likely infeasible. May still be fine to keep the test for illustrative purposes, but if this ever gets fixed on the signalr side, this test won't let you know 99%.

this is also true. safe to throw away my original comment. still valuable to have this test next to the working case to try and explain the issue, i think.